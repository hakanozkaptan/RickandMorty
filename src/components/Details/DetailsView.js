import React, { Component } from 'react';
import { connect } from 'react-redux';
import LazyLoad from 'react-lazy-load';
import { Link } from 'react-router-dom';
import moment from 'moment';
import agent from '../../agent';

import {
  CHARACTERS_PAGE_LOADED,
  CHARACTERS_PAGE_UNLOADED,
  CHARACTERS_EPISODE_LOADED,
  CHARACTERS_EPISODE_UNLOADED
} from '../../constants/actionTypes';

import noImage from '../../assets/images/noimage.jpg';

const mapStateToProps = state => ({
  ...state.character,
});

const mapDispatchToProps = dispatch => ({
  onLoad: payload =>
    dispatch({ type: CHARACTERS_PAGE_LOADED, payload }),
  onUnload: () =>
    dispatch({ type: CHARACTERS_PAGE_UNLOADED }),
  onEpisodeLoad: payload =>
    dispatch({ type: CHARACTERS_EPISODE_LOADED, payload }),
  onEpisodeUnload: () =>
    dispatch({ type: CHARACTERS_EPISODE_UNLOADED })
});

class DetailsView extends Component {
  componentWillMount() {
    const id = this.props.match.params.id;

    const loadCharacterPromise = Promise.all([agent.Characters.getCharacterDetail(id)]);
    loadCharacterPromise.then((character) => {
      const promises = character[0].episode.map((episode) => {
        return agent.Episode.get(episode);
      });

      this.props.onEpisodeLoad(Promise.all(promises));
    });

    this.props.onLoad(loadCharacterPromise);
  }

  componentWillUnmount() {
    this.props.onUnload();
    this.props.onEpisodeUnload();
  }

  createCharacterModel = (character) => {
    const model = {
      name: character.name ? character.name : '-',
      image: character.image ? character.image : noImage,
      location: character.location ? character.location.name : 'Unknown',
      formattedCreateDate: character.created ? moment(character.created).fromNow() : '-'
    }

    return model;
  }

  render() {
    const { character, episodes } = this.props;

    if (!character) {
      return null;
    }

    const characterModel = this.createCharacterModel(character);

    return (
      <div className="details-container">
        <div className="button-center">
          <Link to="/"><div className="bk-btn"><div className="bk-btn-triangle"></div><div className="bk-btn-bar"></div></div>
          </Link>
        </div>
        <div className="card">
          <LazyLoad height={300} offsetVertical={300}>
            <img className="card-img-top" src={characterModel.image} alt={characterModel.name} />
          </LazyLoad>
          <div className="filler" />
          <div className="card-body">
            <h4 className="card-title">{characterModel.name}</h4>
            <p className="location">
              <label>Location: </label>
              {characterModel.location}
            </p>
            <p>
              <label>Created Date: </label>
              {characterModel.formattedCreateDate}
            </p>
          </div>
          <div className="card-footer">
            {episodes.length > 0 && (
              <div className="episode">
                <ul>
                  <li><label>Episodes</label></li>
                  {episodes.map((episode, index) => {
                    return (
                      <li key={'episode_' + episode.id}><strong>{(index + 1)}</strong> - {episode.name}</li>
                    );
                  })}
                </ul>
              </div>
            )}
          </div>
        </div>
      </div>
    );
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(DetailsView);
