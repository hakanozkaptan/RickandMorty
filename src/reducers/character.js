import {
  CHARACTERS_PAGE_LOADED,
  CHARACTERS_PAGE_UNLOADED,
  CHARACTERS_EPISODE_LOADED,
  CHARACTERS_EPISODE_UNLOADED,
} from '../constants/actionTypes';

const defaultState = {
  character: {},
  episodes: []
};

export default (state = defaultState, action) => {
  switch (action.type) {
    case CHARACTERS_PAGE_LOADED:
      return {
        ...state,
        character: action.payload[0]
      };
    case CHARACTERS_EPISODE_LOADED:
      return {
        ...state,
        episodes: [...state.episodes, ...action.payload]
      };
    case CHARACTERS_PAGE_UNLOADED:
      return {
        ...state,
        character: {}
      };
    case CHARACTERS_EPISODE_UNLOADED:
      return {
        ...state,
        episodes: []
      }
    default:
      return state;
  }
};
