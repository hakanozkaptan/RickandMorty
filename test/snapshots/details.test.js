jest.mock('react-redux', () => require('react-redux-mock'));

import React from 'react';
import { MemoryRouter } from 'react-router';
import { __setState } from 'react-redux';
import DetailsView from "../../src/components/Details/DetailsView";
import Enzyme, { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import character from '../responses/character';
import episode from '../responses/episode';

Enzyme.configure({ adapter: new Adapter() });

const characterId = 4;
const match = {
    params: {
        id: characterId
    }
};

describe('detail snapshot testing', () => {
    it('should render correctly if not loaded', () => {
        __setState({
            character: {},
            episodes: []
        });
        
        const componentRender = render(
            <MemoryRouter>
                <DetailsView match={match} />
            </MemoryRouter>
        );

        expect(toJson(componentRender)).toMatchSnapshot();
    });

    it('should render correctly if loaded', () => {
        __setState({
            match: {
                params: {
                    id: characterId
                }
            },
            character,
            episodes: [episode]
        });

        const componentRender = render(
            <MemoryRouter>
                <DetailsView match={match} />
            </MemoryRouter>
        );

        expect(toJson(componentRender)).toMatchSnapshot();
    });
});