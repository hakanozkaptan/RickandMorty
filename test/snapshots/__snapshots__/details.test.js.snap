// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`detail snapshot testing should render correctly if loaded 1`] = `null`;

exports[`detail snapshot testing should render correctly if not loaded 1`] = `null`;
